inference-engine/thirdparty/clDNN/src/gpu/loop_gpu.cpp

```diff
@@ -0,0 +1,399 @@
+            break;
+        }
+        case data_types::i32: {
+            mem_lock<int32_t> lock_prim_output{mem};
+            trip_count = *lock_prim_output.data();
+            break;
+        }
+        case data_types::i64: {
+            mem_lock<int64_t> lock_prim_output{mem};
+            trip_count = *lock_prim_output.data();
+            break;
+        }
+        default:
+            assert(false);
+        }
+        return trip_count;
+    }
+
+    static void copy_buffer(cldnn::memory_impl& src_mem, cldnn::memory_impl& dst_mem,
+                            const size_t size, const size_t src_offset = 0, const size_t dst_offset = 0) {
+        assert(src_mem.get_layout().data_type == dst_mem.get_layout().data_type);
+
+        size_t bytes_per_element = data_type_traits::size_of(src_mem.get_layout().data_type);
+        mem_lock<uint8_t> from_lock{ src_mem };
+        mem_lock<uint8_t> to_lock{ dst_mem };
+
+        const size_t byte_size_to_copy = size * bytes_per_element;
+        const auto src = from_lock.begin() + src_offset * bytes_per_element;
+        const auto dst = to_lock.begin() + (dst_offset * bytes_per_element);
+        std::copy(src, src + byte_size_to_copy, dst);
+    }
+
+    static void write_scalar_value(memory_impl& mem, int64_t input) {
+        const layout& prim_layout = mem.get_layout();
+
+        switch (prim_layout.data_type) {
+        case data_types::u8: {
+            assert(input >= std::numeric_limits<uint8_t>::min() &&
+                   input <= std::numeric_limits<uint8_t>::max());
+            mem_lock<uint8_t> lock_prim_output{mem};
+            *lock_prim_output.data() = static_cast<uint8_t>(input);
+            break;
+        }
+        case data_types::i8: {
+            assert(input >= std::numeric_limits<int8_t>::min() &&
+                   input <= std::numeric_limits<int8_t>::max());
+            mem_lock<int8_t> lock_prim_output{mem};
+            *lock_prim_output.data() = static_cast<int8_t>(input);
+            break;
+        }
+        case data_types::i32: {
+            assert(input >= std::numeric_limits<int32_t>::min() &&
+                   input <= std::numeric_limits<int32_t>::max());
+            mem_lock<int32_t> lock_prim_output{mem};
+            *lock_prim_output.data() = static_cast<int32_t>(input);
+            break;
+        }
+        case data_types::i64: {
+            mem_lock<int64_t> lock_prim_output{mem};
+            *lock_prim_output.data() = input;
+            break;
+        }
+        default:
+            assert(false);
+        }
+    }
+
+    event_impl::ptr execute_impl(const std::vector<event_impl::ptr>& events, loop_inst& instance) override {
+        for (auto& e : events)
+            e->wait();
+        auto& outer_network = instance.get_network();
+        const uint32_t& net_id = instance.get_network().get_id();
+        auto ev = outer_network.get_engine().create_user_event(net_id, false);
```

Comment:
Why do you create user event for synchronization? Can't you just aggregate events from the last iteration?

Reply:
We need to ensure that body network is executed completely to copy the sliced output memories of the body network to the concatenated outer output memory. I also tried `cl::usm::memcpy` to copy sliced/concatenated memory between the body network and the main network by GPU and set the event properly without creating the user event as you mentioned, but there was no gain.

Reply:
BTW, for `usm::memcpy` I believe @sshlyapn discovered some time ago that it had huge time for small buffers copy. Maybe that why it didn't help to you, and we just need to investigate this memcpy perf issue

@sshlyapn please confirm if that was the case.